### PR TITLE
Allow to import python modules directly / remotePath feature

### PR DESCRIPTION
**Pull request by phorward**
# remotePath-feature for Pyodide

This pull request implements the remotePath-feature. This feature makes Pyodide capable to fetch Python modules remotely from a set of given URLs. Both single Python py-files as well as Python-modules in sub-folders using `__init__.py` are supported.

In the demonstrated example below, there is a module called `xyz` located at https://phorward.info/tmp/pyodide/xyz/, which in turn imports the modules x and y. 

Python console:
python-console

JavaScript console:
image

Any remotely fetched module is parsed for further imports which are imported as well, including pre-build packages from Pyodide. Due to this entirely changed behavior on how imports are resolved,  packages from the pyodide packages.json take a higher precedence than remotely loaded packages.

The feature is already tested well in a re-implementation of a Python-based HTML5-web-app that previously was compiled into JavaScript using PyJS, but now can directly be interpreted ([further details here](https://www.viur.is/blog/pyodide))

---
[INITIAL MESSAGE OF THIS PULL REQUEST]
Hi there!

This is a work-in-progress draft to resolve my problem I reported in #481 to allow importing Python files directly, which are not pre-packaged with pyodide, but directly served. So this pull request now makes it possible to make a 
```python
import a
```
and `a.py` is fetched from the same source as pyodide is served from (baseURL). In case `a.py` imports further modules, let it be `b.py` and `c.py`, these are also fetched. This is done until all required packages are fetched, including imports to available packages, as it is the case before.

Python files fetched from the server are stored into the browser cache using [FS.writeFile](https://emscripten.org/docs/api_reference/Filesystem-API.html#FS.writeFile), and then imported by Python by setting `PYTHONPATH` to `/`. I hope this won't be a security problem.

A demo of this feature is available here: https://phorward.info/tmp/pyodide/
-  [abc.html](https://phorward.info/tmp/pyodide/abc.html) does the above import of a.py, b.py, c.py and the package html5
- [game.html](https://phorward.info/tmp/pyodide/game.html) just does an `import game`, loads game.py and html5 and executes.
- [console.html](https://phorward.info/tmp/pyodide/console.html) can also be used: just type `import game` or `import a` there to see what's happening (take a look into the JavaScript console for output)

This is just a first draft, I want to enhance it also to import module directories served. 
Is this useful also for others? Does anyone have suggestions for improvements?

### CODE REVIEW

**Comment by phorward** [comment]
Hello again,

I've recently updated my pull request and implemented directory fetching also. Now, a pure Python source module can be placed as directory with an `__init__.py` and fetched from there.

Enhanced demos (for results, look, again, into the JS console!)
- [xyz.html](https://phorward.info/tmp/pyodide/xyz.html)
- [xyz2.html](https://phorward.info/tmp/pyodide/xyz2.html)

Alternatively, open [console.html](https://phorward.info/tmp/pyodide/console.html) and type
```
import xyz2
xyz2.sayHello(100)
```
and see what happens! So nice, that was the feature I've needed with #481. I will do further tests now with real Python software for more testing.

**Comment by mdboom** [comment]
Awesome contribution, thanks!

Just a high-level braindump, and then I'll try to do a proper code review after you declare it ready and remove the `WIP`.  (As an aside, please format the code consistent with the rest -- there's a check in `make lint` that uses `clang-format`).

On first blush, my biggest concern is that loading files from the same place as the Pyodide distribution is often, but not always, the desirable thing.  A user may want to use Pyodide from the "official" CDN, for example, but then load their custom code from elsewhere.  Python `import` statements, of course, wouldn't allow for a full URL.  But there's an analogy to be found, perhaps, with how `sys.path` is used to specify places on the user's filesystem where modules are loaded from.  We could add, for example, `sys.remote_path` (or `pyodide.remote_path`) that is a list of base URLs where remote modules would be loaded from.  It's not a perfect idea, because searching a bunch of bases, and I guess using 4XX error codes to determine when they are missing would be kind of expensive and chatty on the network.  If there's only one entry in `remote_path`, though, it's maybe not so bad.

I think there's also a use case where you want to load a module from a fully-specified URL.  Again, the `import` syntax won't support that, but a `pyodide.import('https://example.com/module.py')` could work (and could be a follow-on PR to this since all the hard parts to make that possible would also be required for this...)

Hope that's useful feedback!  Overall, I think this is a great idea and fills an important use case hole.

**Comment by mdboom** [comment]
One other thing -- I think we should make this explicit in the import somehow.  If you use this to load a module called `foo` and we later add a module `foo` to Pyodide, we'll break people's code.  Requiring `from . import foo` for this functionality might work.

**Comment by phorward** [comment]
Hi @mdboom, 
thank you very much for your first comments and concerns on this issue.
First of all, I've fixed the linting errors so the CI should be working again.

Your concerns on how to integrate this new feature into pyodide are absolutely valid. I think, that both, making the fetch source URLs configurable via `sys.remote_path` or `pyodide.remote_path` (I would prefer latter one) and providing a `pyodide.import()`-function as well are good ideas.

In general, it would be possible to control the entire Python source remote fetching behavior (do you have a good idea how to name this feature in future?) using `pyodide.remote_path`:

- `pyodide.remote_path = ["."]` would allow for fetching from the current pyodide baseURL,
- `pyodide.remote_path = ["https://alpha.iodide.io/pyodide"]` would allow for fetching only from https://alpha.iodide.io/pyodide,
- `pyodide.remote_path = ["https://alpha.iodide.io/pyodide", "."]` would allow for fetching first from https://alpha.iodide.io/pyodide and, when not found, from current pyodide baseURL,
- `pyodide.remote_path = None` could be set as the default and entirely turns the feature off, so we have the same behavior of the `import` statement like before.

This does also bypass the problem with future modules; If the users want to import Python modules from elsewhere, they have to turn the feature on and deal with the consequence, that a fetched package takes precedence over a build-in one.

I think it's best when I implement your ideas in some way that we can discuss about, when I find some time. Furthermore, I'm not sure about any security vulnerability this feature might open, and how to prevent them. I think, turning the feature entirely off except when wanted (as described above) is a good solution for its introduction.

Thanks again, and hopefully this feature finds more friends!

**Comment by mdboom** [comment]
I agree 100% with your proposal above.  Thanks!

**Comment by phorward** [comment]
Hi @mdboom,

I think I've made it with the last commit. There's also a section now in the docs describing the new feature, you can preview it [here](https://github.com/phorward/pyodide/blob/feature/import-python-files/docs/api_reference.md#pyodideremotepath).

I've tested it successfully with my test setup on https://phorward.info/tmp/pyodide/, but had to do some CORS header configurations. It can load remote contents perfectly from localhost also.

Using this version I now want to do serious development with it, and prove its stability and usability.

**Comment by phorward** [comment]
Haha, first bug fixed.

But I also found another problem: Modules are fetched multiple times because of a missing "I'm already fetching this"-cache. This is some more work to do and I'm on it, but it doesn't change the general functionality.

**Comment by mdboom** [comment]
Thanks, @phorward.  Really cool to see progress on this.  I should probably have some bandwidth next week to look at this and kick the tires.

**Comment by phorward** [comment]
Hi @mdboom, 

I just wanted to report that I finished all the stuff I requested for this feature now. It works really great: I was able to port a (yet) Python2-only web-app that was compiled into JavaScript using PyJS before, now into a Python3-pyodide-powered web-app, in just a couple of hours. It runs amazingly fast and the remotePath feature works very great, and is ideal for testing and debugging.

I think this can be merged now, if you're comfortable with it. One thing I recognized is, when build-in Python modules are imported, the remotePath feature also tries to fetch them (e.g. `os.py`, `json.py` when `import os, json` is given), and raises some 404s. Python itself also prefers to import local file modules first, when available, and then falls back to the build-in modules, so this behavior is in my opinion OK and normal. If not, we might think about a "whitelist" that can be configured to ignore some files.

File: src/main.c [hunk 1]
```diff
@@ -16,6 +16,7 @@ main(int argc, char** argv)
   hiwire_setup();
 
   setenv("PYTHONHOME", "/", 0);
+  setenv("PYTHONPATH", "/", 0);
```
[the remark this answers is not in the document]
Comment (phorward) [reply src/main.c]: Hi, I've made it static now and set "/usr/lib/python3.7/remote-packages" - I hope that is ok?

File: docs/api_reference.md [hunk 2]
```diff
@@ -141,6 +141,26 @@ This is equivalent to calling `repr(obj)` in Python.
 |------------|---------|-------------------------------------------|
 | *str_repr* | String  | String representation of the input object |
 
+### pyodide.remotePath
```
[the remark this answers is not in the document]
Comment (phorward) [reply docs/api_reference.md]: I'm not sure how you mean this exactly. I found out, that it's possible to change `pyodide.remotePath` also on the Pyodide console.html, see here:

```
Welcome to the Pyodide terminal emulator 🐍
>>> pyodide.remotePath
/
>>> pyodide.remotePath = [] #disable!
>>> import xyz2
Error: Traceback (most recent call last):
  File "/lib/python3.7/site-packages/pyodide.py", line 42, in eval_code
    exec(compile(mod, '<exec>', mode='exec'), ns, ns)
  File "<exec>", line 1, in <module>
ModuleNotFoundError: No module named 'xyz2'
>>> pyodide.remotePath = ["/"] #enable!
>>> import xyz2
>>> xyz2.sayHello() #works!
>>>
```

**Comment by phorward** [comment]
> This is really turning into a very nice feature!

Thanks! :+1: I'm very happy that Pyodide finally breaks a problem that we needed to solve for a long time, and it does its job very well - thanks for making this happen!

> I just gave it a first read through -- hope to have a more thorough look again soon.

That was very helpful. I changed almost all of your request and hope you're ok with them.

> It does seem that all of the unit tests are broken -- it's not able to load the `test` module to get the CPython unit tests. Something about this is inhibiting that, I guess. Are you able to run the tests locally and diagnose?

I've checked it out but was yet unable to find out the reason. E.g. with micropip, it says "firefox_micropip" fails, but when I do `import micropip` on Pyodide console there's no problem at all. I'm also not familiar with test suites like this, yet. Can you give me a hint on that?

File: src/main.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (phorward) [reply src/main.c]: we can also make a #define somewhere?

**Comment by UnHumbleBen** [comment]
Any update? I would really like to use this feature.

**Comment by phorward** [comment]
Hi @mdboom and @UnHumbleBen,

thanks for asking.

Yes, I did several improvements and a general code clean-up (mostly dead code, resulting from older tests) of the remotePath-feature within the last commits. Unfortunately, I was not able to get the entire test suite back to work without errors. This is, because the new way of how modules are imported into pyodide breaks some tests that need to be rewritten or reassessed.

Anyway, most of the tests are succeeding now due a general fix that had its origin in the mapping between package names and their import names (e.g. "beautifulsoup4" and "bs4"). [_resolveImports() in pyodide.js](https://github.com/phorward/pyodide/blob/feature/import-python-files/src/pyodide.js#L101) now allows for both package namings, the Python import name (e.g. bs4) and the package name (e.g. beautifulsoup4). In Pyodide's current master, this mapping is handled directly in runpython.c so loadPackage from pyodide.js just works with the package names. But the entire import handling with this feature is now handled in _resolveImports from pyodide.js to get the remotePath behavior working, also in case that a remotely imported module imports a package. I hope this solution is okay, but maybe you have a better idea where to draw the line between the import names and the package names.

Currently, Pyodide still throw some errors when remotely imported modules that are importing build-in modules like "os", because it is first looked for `os.py` and `__init__/os.py`. For this case, I want to introduce a whitelist of import names that are silently dropped - I would like to extract this list from [stdlib_list](https://pypi.org/project/stdlib-list/) and locally store it in pyodide.js as first attempt. Later on, this can be improved.

Hopefully @mdboom will take a look on this again soon. Maybe there are still some issues that need to be improved, but in general this feature works already very well.

I'll post another update on this as soon as I've implemented the whitelist and the feature runs well.

**Comment by phorward** [comment]
Hi @UnHumbleBen,
I've now updated the Pyodide build on my test installation at https://phorward.info/tmp/pyodide/ to serve a current image build using docker.

It runs very well with `import xyz`, see the two screenshots below, one is from the Pyodide python console (https://phorward.info/tmp/pyodide/console.html), and another is directly from a JavaScript console.

Python console:
python-console

JavaScript console
image

**Comment by phorward** [comment]
You can download the build Pyodide version with remotePath-feature from [here](https://phorward.info/tmp/pyodide/builds/) now. You have to fix the baseURL in pyodide_dev.js when  Pyodide is served from a sub-folder on a web-server or application.

**Comment by phorward** [comment]
> What do you mean by "fix the baseURL"?

I think your problem is exactly what I meant: Both `pyodide.js` and `pyodide_dev.js` contain a line

```javascript
 // this is the line in pyodide.js
var baseURL = self.languagePluginUrl || 'https://pyodide.cdn.iodide.io/';

 // this is the line in pyodide_dev.js
var baseURL = self.languagePluginUrl || '';
```

When you use pyodide.js, it will load the most recent official pyodide build from pyodide.cdn.iodide.io. Switching to pyodide_dev.js will use the currently served root of localhost (or wherever you serve it).
In my live demo installation, I needed to change this line (in this case in pyodide_dev.js) to read
```javascript
var baseURL = 'https://phorward.info/tmp/pyodide/';
```
so that it works.

I've just tested it with the default pyodide.js from my build-tarball and got the same results as you, so changing the baseURL variable there to read just

```javascript
var baseURL = '';
```

makes it to work as expected.

I'm currently not sure how this variable is explicitly handled, so @mdboom has to give us a statement on this. (there is also a TODO-comment above the line that this must be revised at some time). In case you're facing further problems with this feature, please contact me directly by mail because this should not be discussed as part of the pull request.

**Comment by phorward** [comment]
> You can also set self.languagePluginUrl before loading Pyodide so you don't need to hack the file.

Thanks, I was able to get it to work. Here's a complete example for @UnHumbleBen that should work out of the box:
```html
<!doctype html>
<html>
  <head>
    <meta charset="UTF-8">
    <script>
      window.languagePluginUrl = "./";
    </script>
    <script src="./pyodide.js"></script>
  </head>
  <body>
    <script>
      languagePluginLoader.then(() => {
        pyodide.remotePath = ["/"];
        pyodide.runPythonAsync("import xyz\nxyz.sayHello()");
      });
    </script>
  </body>
</html>
```

Theres an updated, pre-build pyodide [here](https://phorward.info/tmp/pyodide/builds/pyodide-remotePath-build-20190828.tar.gz), https://phorward.info/tmp/pyodide/ was also updated. The Pyodide console can be found [here](https://phorward.info/tmp/pyodide/remotePath-console.html) now.

I think that I've finally made it now and was able to run the complete (yet existing) test suite without problems. Anyway, the test suite currently does not cover the remotePath-feature itself, I will deliver this soon.

**Comment by phorward** [comment]
Hi there,

I've implemented some tests for the remotePath feature now. So far, they look good on my local machine's docker instance.
```
test/test_remotepath.py::test_runpythonasync_remotepath_single[firefox] PASSED [ 95%]
test/test_remotepath.py::test_runpythonasync_remotepath_folder[firefox] PASSED [ 95%]
test/test_remotepath.py::test_runpythonasync_remotepath_xyz[firefox] PASSED [ 95%]
test/test_remotepath.py::test_runpythonasync_remotepath_mixedimports[firefox] PASSED [ 95%]
...
test/test_remotepath.py::test_runpythonasync_remotepath_single[chrome] PASSED [ 95%]
test/test_remotepath.py::test_runpythonasync_remotepath_folder[chrome] PASSED [ 95%]
test/test_remotepath.py::test_runpythonasync_remotepath_xyz[chrome] PASSED [ 95%]
test/test_remotepath.py::test_runpythonasync_remotepath_mixedimports[chrome] PASSED [ 95%]
```
Unfortunately, the last build faced a memory error during the chromium build (last run: Message: unknown error: memory access out of bounds) and this at a test which don't have to do with remotePath. All firefox tests succeeded.

Any idea how to deal with this?

**Comment by mdboom** [comment]
The Chrome tests are a bit flaky and this kind of thing isn't uncommon.  I usually don't worry about these class of failures (since I haven't found a good workaround for them, yet).

**Comment by phorward** [comment]
Hi there, this is a status update on this pull requests.

Since the last commit b29de212656293c0021175e6683d882b0cadfcbc, this feature runs very fine for our demands. To limit the number of 404s due to examining package files, a whitelist was added (initially in 4da6396e86e198f5acc1b7f4dc2031fb60558fe8) for common Python build-in packages. I'm also planning to add some kind of "mapping" feature here so only Python files reading from a map are loaded using the remotePath feature.

I've already ported another web-app to Pyodide successfully, but using [our fork](https://github.com/mausbrand/pyodide) that integrates pull requests #525 and #512, and comes without any of the packages delivered with Pyodide (we just need the raw Python interpreter for our use-case).

I hope this feature could also be useful for others. Otherwise, we need to continue with further improvements on our fork now, and merge improvements made here in the upstream into this, in future.

**Comment by mdboom** [comment]
Thanks for the update.  I'm going to give it a final review now.

**Comment by phorward** [comment]
Hi there, this is a status update on this pull request.

Latest considerations about our use-cases now moved away from the attempts started here. To minimize the number of 404 requests that are automatically arise when the remotePath feature tries to find pathes, we built some tiny tools that just build static file lists and loading the package files from there.

Using this new attempt, we resulted in a great tool chain now which is capable to use the official Pyodide binary distributions from [here](https://github.com/iodide-project/pyodide/releases) without any modification.

For those who are interested:
- [`get-pyodide.py`](https://github.com/viur-framework/viur-vi/blob/master/get-pyodide.py) downloads and extracts a minimal Pyodide distribution
- [`gen-files-json.py`](https://github.com/viur-framework/viur-vi/blob/master/gen-files-json.py) generates a JSON file with the files to be loaded, and [some piece of JavaScript code](https://github.com/viur-framework/viur-vi/blob/master/public/js/app.js) finally installs these files in the Emscripten filesystem environment so that Pyodide can find them.

In case this pull request is not necessary anymore, I would recommend to close it for now. It was a nice exercise to me and finally results in this good toolchain with minor dependencies. I wouldn't continue this attempt anymore for now.

Thanks for any advise!

**Comment by pannous** [comment]
Eagerly awaiting for this feature to be merged!
Anything we can do to assist?

**Comment by DonJayamanne** [comment]
Thanks for this great feature.
I've found that it doesn't work with `runPython` anymore.
Basically the following code and similar ones fail
```python
import contextlib
contextlib.contextmanager
# contextmanager doesn't exist
```

Please note, i'm using the build from here https://github.com/iodide-project/pyodide/pull/489#issuecomment-525822045

**Comment by DonJayamanne** [comment]
@phorward 
```
For those who are interested:
get-pyodide.py  downloads and extracts a minimal Pyodide distribution
```

Thanks, however it looks like these files are no longer available. Any chance you can share these (if you still have them).
Thanks

**Comment by phorward** [comment]
> Hope you can share this. Thanks.

Sure, I'm sorry the link is gone. All you need can be found in the repo https://github.com/mausbrand/pyodide-app-stub. Our company is already using our Pyodide-powered WebApp in production but for a private customer's project.

**Comment by DonJayamanne** [comment]
Thanks, wish they was an easier way to build the js files from python packages, can't get this working on my machine 😔

**Comment by phorward** [comment]
> Thanks, wish they was an easier way to build the js files from python packages, can't get this working on my machine pensive

What doesn't work? A little more information would be helpful. The pyodide-app-stub example normally runs out of the box, you don't need to creeate packages.

Anyway, when you want to build native Pyodide packages, use the Docker container from the repo. It runs best and you don't have to hazzle with your local installation. I got Pyodide natively compiled on my Arch Linux box with a few modifications but it isn't worth it.

**Comment by phorward** [comment]
Hi, I was sure I've closed this pull request due substitution of another solution a while ago, now its open again.

**Comment by Emmanuel12345678** [comment]
I keep getting module not found error

**Comment by phorward** [comment]
> I keep getting module not found error when using your pyodide.remotePath

Hi, this feature was abandoned. See https://github.com/mausbrand/flare/blob/master/assets/js/init.js for a proven and well working solution.

**Comment by Emmanuel12345678** [comment]
Please, how do I use the init.js to import modules from serverside?

**Comment by phorward** [comment]
Hi, I'm sorry the init.js is only half way, how to use it can be seen here: https://github.com/mausbrand/flare/blob/master/hello.html Please do also checkout the ReadMe. A module is a directory with Python Files and __init__.py fetched from somewhere, using a files.json file that must be available as index. Its more manual than this pull request was, but also running faster. Please do also checkout the repository pyodide-app-stub linked above.

**Pull request closed without merging**